General use page (#15)
Luos-io/doc: src/_pages/first_steps/general-use.md

## Node basic capacities
Nodes can have capacities such as measuring the core temperature, sending the processor unique ID, or input voltage. Nodes capacities are commonly shared with all modules hosted in. Any module hosted in a node is able to send its input voltage.

<a name="module"></a>
## Module
A module is a block of code which is able to communicate with any other modules in the Luos network. Each module provides a particular set of tasks such as managing a motor, handling a laser range finder, or compute an inverse-kinematics, for example.
**Each module is hosted in a single node**, but a node can handle several modules at the same time and manage communication between them and between other modules hosted in other nodes, using the same interface.

For example, the [Dynamixel board]({{boards_path}}dxl.md) provided by Luos Robotics can dynamically create and manage Dynamixel modules depending on the number of Dynamixel motors linked to it. Any Dynamixel module can get or set values to other Dynamixel modules on the same node or to any other module in any other node of the network.

> <a href="/_assets/img/feature-module-node-board.jpg" target="_blank"><img src="/_assets/img/feature-module-node-board.jpg" width="800px" /></a>

Reviewer: This one still opened and seems to continue to fail.
Do I have to make my review anyway?

> <a href="/_assets/img/feature-module-node-board.jpg" target="_blank"><img src="/_assets/img/feature-module-node-board.jpg" width="800px" /></a>

Author: I forgot this one sorry. Let me add the picture link

## Module basic properties

> To properly work, each module owns some properties allowing to other modules to recognize and access it:

Reviewer: ```suggestion
Modules owns some properties allowing to other modules to recognize and access it:
```

| Name | Description | Format |
| :---: | :---: | :---: |

> | **ID** | The ID is an unique number given to each module depending on their physical position. The system automatically assign each ID during the detection phase. If you move a module from a microcontroller A to a microcontroller B on a given robot, the ID will change. In the same way, if you change the wiring order of a microcontroler on the network on a given robot, the ID will change too. | Integer<br />e.g. `Id=1` |

Reviewer: ```suggestion
| **ID** | The ID is an unique number given to each module depending on their physical position. The system automatically assign each ID during the detection phase. If you change the wiring order of nodes on the network, the IDs of your modules will change. | Unsigned Integer 12 bits<br />e.g. `Id=1` |
```

> | **TYPE** | The type defines the module purpose. A few types are predefined and can be used, or new ones can be created. The module type cannot be changed after module initialization. | String<br />e.g. `type=DISTANCE_MOD` |

Reviewer: ```suggestion
| **TYPE** | The type defines the module purpose. A few types are predefined and can be used, and new ones can be created. The module type cannot be changed at run time. | Enum Unsigned Integer 12 bits<br />e.g. `type=DISTANCE_MOD` |
```

> | **ALIAS** | Alias is the name of the module. It is used to easily identify a module. Each module has a **default alias** who can be changed by users. For example, a module with the default alias `motor_mod` can be named `left_knee_motor` by user. This new name will be stored in the non-volatile memory of the board. As we do not want to have multiple modules with the same name, a duplicate name on your system will be automatically assigned with an incrementing number at its end, in the network. You can go back to the default name by setting a void name (`""`) to a module. | String<br />e.g. `alias="gate"` |

Reviewer: ```suggestion
| **ALIAS** | Alias is the name of the module. It is used to easily identify a module. Each module has a **default alias** who can be changed by users. For example, a module with the default alias `motor_mod` can be named `left_knee_motor` by user. This new name will be stored in the non-volatile memory of the board. As we do not want to have multiple modules with the same name, a duplicate name on your system will be automatically assigned with an incrementing number at its end, in the network. You can go back to the default name by setting a void name (`""`) to a module. | String 16 bytes<br />e.g. `alias="gate"` |
```

## Routing table
A routing table is a "service" managed by the Luos network and available for any module in any node. This service lists all the modules and allow to any module to get and use basic information of any other modules. The routing table's data can be loaded or auto-generated during detection.

[Go to Routing table page](/_pages/modules/routing-table.md).

## Module detection

> The module detection assigns IDs to modules depending on their node's physical position on the cabling, and generates a route table.

Reviewer: ```suggestion
The module detection assigns IDs to modules depending on their node's physical position on the network, and generates a route table.
```

IDs are assigned from the nearest to the furthest node branch by branch, from the point of view of the module running the detection. Following this logic, the module running the detection will have the ID 1, the next one will have the ID 2, etc.

Multiple detection by different modules at the same time is not allowed.

> It is possible to detect the network frequently to discover included or excluded modules at run time.

Reviewer: ```suggestion
It is possible to detect the network frequently to dynamically discover included or excluded modules at run time.
```

> ## Module types

Reviewer: ```suggestion
## Modules
```

Modules are code sections which enclosed several features. They are embedded into nodes.
There are two types of modules:

### Drivers

> Drivers are modules managing hardware. They are passive modules that retrieve pieces of information from electronic devices.

Reviewer: ```suggestion
Drivers are modules managing hardware. They are passive modules that retrieve pieces of information from electronic devices, format it, and share it trough the Luos API.
```
